art/attacks/inference/membership_inference.py

```diff
@@ -0,0 +1,346 @@
+
+    def infer(self, x: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
+        """
+        Infer membership in the training set of the target estimator.
+
+        :param x: Input records to attack.
+        :param y: True labels for `x`.
+        :return: An array holding the inferred membership status, 1 indicates a member and 0 indicates non-member.
+        """
+        if self.estimator.input_shape[0] != x.shape[1]:
+            raise ValueError("Shape of x does not match input_shape of classifier")
+
+        y = check_and_transform_label_format(y, len(np.unique(y)), return_one_hot=True)
+
+        if y.shape[0] != x.shape[0]:
+            raise ValueError("Number of rows in x and y do not match")
+
+        if self.input_type == 'prediction':
+            features = self.estimator.predict(x).astype(np.float32)
+        elif self.input_type == 'loss':
+            features = self.estimator.loss(x, y).astype(np.float32).reshape(-1, 1)
+
+        if self.default_model and self.attack_model_type == 'nn':
+            self.attack_model.eval()
+            inferred = None
+            test_set = AttackDataset(x1=features, x2=y)
+            test_loader = DataLoader(test_set, batch_size=self.bs, shuffle=True, num_workers=0)
+            for input1, input2, targets in test_loader:
+                outputs = self.attack_model(input1, input2)
+                predicted = torch.round(outputs)
+                if inferred is None:
+                    inferred = predicted.detach().numpy()
+                else:
+                    inferred = np.vstack((inferred, predicted.detach().numpy()))
+        else:
+            inferred = np.array([np.argmax(arr) for arr in self.attack_model.predict(np.c_[features, y])])
+        return inferred
+
+
+class MembershipInferenceWhiteBoxNeuralNetwork(InferenceAttack):
+    """
+        Implementation of a learned white-box membership inference attack.
+
+        This implementation can use as input to the learning process activations and/or gradients of one or more layers,
+        depending on the provided configuration.
+    """
+    def __init__(self, classifier: Classifier):
+        raise NotImplementedError
+
+    def infer(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> np.ndarray:
+        """
+        Infer sensitive properties (attributes, membership training records) from the targeted estimator. This method
+        should be overridden by all concrete inference attack implementations.
+
+        :param x: An array with reference inputs to be used in the attack.
+        :param y: Labels for `x`. This parameter is only used by some of the attacks.
+        :return: An array holding the inferred properties.
+        """
+        raise NotImplementedError
+        # if NeuralNetworkMixin in type(self.estimator).__mro__:
+        #     activations = self.estimator.get_activations(x, self.estimator.layer_names[-1], batch_size=self.bs)
```

Comment:
How is this class used?

Reply:
It is not at the moment. It's just a placeholder, I didn't have time to implement it for this release. I guess I should remove it then, right?

Reply:
Yes, removing was my first thought because it does not implement methods. But if you would like to claim the name space and have plans to implement it in the very near future you can leave it, in that case I would ask to add a comment describing the plans.